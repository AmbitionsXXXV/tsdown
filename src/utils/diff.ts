import { diffLines } from 'diff'
import pc from 'picocolors'

export function diff(original: string, updated: string): string {
  let text = ''
  const diff = diffLines(original, updated, {
    ignoreWhitespace: true,
    ignoreNewlineAtEof: true,
  })
  for (const line of diff) {
    const { value } = line
    text += line.added
      ? pc.green(value)
      : line.removed
        ? pc.red(value)
        : pc.gray(line.value)
  }
  return text
}
