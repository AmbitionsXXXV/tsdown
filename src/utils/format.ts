export function formatBytes(bytes: number): string {
  if (bytes === Infinity) {
    return 'too large'
  }
  const numberFormatter = new Intl.NumberFormat('en', {
    maximumFractionDigits: 2,
    minimumFractionDigits: 2,
  })

  return `${numberFormatter.format(bytes / 1000)} kB`
}
