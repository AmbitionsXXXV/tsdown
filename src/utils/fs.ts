import { access } from 'node:fs/promises'

export function fsExists(path: string): Promise<boolean> {
  return access(path).then(
    () => true,
    () => false,
  )
}
