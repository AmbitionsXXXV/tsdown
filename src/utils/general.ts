export function toArray<T>(
  val: T | T[] | null | undefined,
  defaultValue?: T,
): T[] {
  if (Array.isArray(val)) {
    return val
  } else if (val == null) {
    if (defaultValue) return [defaultValue]
    return []
  } else {
    return [val]
  }
}
