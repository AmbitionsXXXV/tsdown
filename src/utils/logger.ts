import process from 'node:process'
import { consola, type ConsolaInstance } from 'consola'
import Debug from 'debug'

/**
 * Logger instance
 */
export const logger: ConsolaInstance = consola.withTag('tsdown')
export const debug: Debug.Debugger = Debug('tsdown')

export function setSilent(silent: boolean): void {
  if (!('CONSOLA_LEVEL' in process.env)) {
    logger.level = silent
      ? 0 // Fatal and Error
      : 3 // Informational logs, success, fail, ready, start, ...
  }
}
