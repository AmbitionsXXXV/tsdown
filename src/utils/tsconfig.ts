import { up as findUp } from 'empathic/find'

export function findTsconfig(
  cwd?: string,
  name: string = 'tsconfig.json',
): string | false {
  return findUp(name, { cwd }) || false
}
