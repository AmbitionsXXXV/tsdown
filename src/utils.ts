import { existsSync } from 'node:fs'
import { unlink } from 'node:fs/promises'
import { globby } from 'globby'
import { type ConsolaInstance, consola } from 'consola'

export async function removeFiles(
  patterns: string[],
  dir: string,
): Promise<void> {
  const files = await globby(patterns, {
    cwd: dir,
    absolute: true,
  })
  await Promise.all(files.map((file) => existsSync(file) && unlink(file)))
}

export const logger: ConsolaInstance = consola.withTag('tsdown')
