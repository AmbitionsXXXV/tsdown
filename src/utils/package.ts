import path from 'node:path'
import { type PackageJson, readPackageJSON } from 'pkg-types'
import { fsExists } from './fs'

export async function readPackageJson(
  dir: string,
): Promise<PackageJson | undefined> {
  const packageJsonPath = path.join(dir, 'package.json')
  const exists = await fsExists(packageJsonPath)
  if (!exists) return
  return readPackageJSON(packageJsonPath)
}

export function getPackageType(pkg: any): 'module' | 'commonjs' {
  if (pkg.type) {
    if (!['module', 'commonjs'].includes(pkg.type)) {
      throw new Error(`Invalid package.json type: ${pkg.type}`)
    }
    return pkg.type
  }
  return 'commonjs'
}
