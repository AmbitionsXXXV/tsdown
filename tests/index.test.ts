import { rm } from 'node:fs/promises'
import { beforeEach, expect, test } from 'vitest'
import { getTestDir, testBuild } from './utils'

beforeEach(async () => {
  const dir = getTestDir()
  await rm(dir, { recursive: true }).catch(() => {})
})

test('basic', async () => {
  const content = `console.log("Hello, world!")`
  const { snapshot } = await testBuild({
    'index.ts': content,
  })
  expect(snapshot).contain(content)
})

{
  const files = {
    'index.ts': "export { foo } from './foo'",
    'foo.ts': 'export const foo = 1',
  }
  test('esm import', async () => {
    await testBuild(files)
  })

  test('cjs import', async () => {
    await testBuild(files, {
      format: 'cjs',
    })
  })
}

test('syntax lowering', async () => {
  const { snapshot } = await testBuild(
    { 'index.ts': 'export const foo: number = a?.b?.()' },
    { target: 'es2015' },
  )
  expect(snapshot).not.contain('?.')
})

test('esm shims', async () => {
  await testBuild(
    { 'index.ts': 'export default [__dirname, __filename]' },
    { shims: true },
  )
})

test('cjs shims', async () => {
  await testBuild(
    {
      'index.ts': `
      import.meta.url === require("url").pathToFileURL(__filename).href
      import.meta.filename === __filename
      import.meta.dirname === __dirname`,
    },
    {
      shims: true,
      format: 'cjs',
    },
  )
})

test('entry structure', async () => {
  const files = {
    'src/index.ts': '',
    'src/utils/index.ts': '',
  }
  await testBuild(files, {
    entry: Object.keys(files),
  })
})

test('bundle dts', async () => {
  const files = {
    'src/index.ts': `
    export { str } from './utils/types';
    export { shared } from './utils/shared';
    `,
    'src/utils/types.ts': 'export let str = "hello"',
    'src/utils/shared.ts': 'export let shared = 10',
  }
  await testBuild(files, {
    entry: ['src/index.ts'],
    dts: { extraOutdir: 'types' },
    bundleDts: true,
  })
})
