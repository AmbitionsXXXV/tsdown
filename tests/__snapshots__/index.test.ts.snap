// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
"//#region index.ts
console.log("Hello, world!");

//#endregion"
`;

exports[`cjs import 1`] = `
""use strict";

//#region foo.ts
const foo = 1;

//#endregion
exports.foo = foo"
`;

exports[`esm import 1`] = `
"//#region foo.ts
const foo = 1;

//#endregion
export { foo };"
`;

exports[`syntax lowering 1`] = `
"//#region index.ts
var _a, _a$b;
const foo = (_a = a) === null || _a === void 0 || (_a$b = _a.b) === null || _a$b === void 0 ? void 0 : _a$b.call(_a);

//#endregion
export { foo };"
`;
